### PR TITLE
cmd: avoid using current symlink in InternalToolPath

### PR DESCRIPTION
**Pull request by mvo5**
The previous cmd.InternalToolPath() made assumptions like that
the /snap/core/current symlink is available when the command is
used. However that is not the case if e.g. core is refreshed.

The new code is much simpler and will just take the basename of
/proc/self/exe and use that basedir as the base for the internal
tool.

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -112,32 +115,30 @@ func coreSupportsReExec(corePath string) bool {
...
 
-	// Is the tool we are after present in the core snap?
-	corePath := newCore
-	coreTool := filepath.Join(newCore, "/usr/lib/snapd", tool)
-	if !osutil.FileExists(coreTool) {
-		corePath = oldCore
-		coreTool = filepath.Join(oldCore, "/usr/lib/snapd", tool)
+	// ensure we never use this helper from anything but
+	if !strings.HasSuffix(exe, "/snapd") {
+		panic("InternalToolPath can only be used from snapd")
```
Comment (niemeyer) [inline cmd/cmd.go]: Thanks for being careful here.

File: cmd/cmd.go [hunk 2]
```diff
@@ -112,32 +115,30 @@ func coreSupportsReExec(corePath string) bool {
...
 	if !osutil.GetenvBool(reExecKey, true) {
 		logger.Debugf("re-exec disabled by user")
 		return distroTool
 	}
 
-	// If the distribution doesn't support re-exec or run-from-core then don't do it.
-	if !distroSupportsReExec() {
+	exe, err := osReadlink("/proc/self/exe")
+	if err != nil {
+		logger.Noticef("cannot read /proc/self/exe: %v, using default command", err)
```
Comment (niemeyer) [inline cmd/cmd.go]: "... using tools outside core", perhaps?

File: cmd/cmd.go [hunk 3]
```diff
@@ -112,32 +115,30 @@ func coreSupportsReExec(corePath string) bool {
...
-	// "spiritual" re-exec so use the same environment variable to decide.
+	// mostly useful for tests
 	if !osutil.GetenvBool(reExecKey, true) {
 		logger.Debugf("re-exec disabled by user")
 		return distroTool
 	}
 
-	// If the distribution doesn't support re-exec or run-from-core then don't do it.
-	if !distroSupportsReExec() {
+	exe, err := osReadlink("/proc/self/exe")
```
Comment (niemeyer) [inline cmd/cmd.go]: Can we please have a reminder here explaining why we're not just using the straight path, to make sure we won't revert into the bug by mistake?

File: cmd/cmd.go [hunk 4]
```diff
@@ -112,32 +115,30 @@ func coreSupportsReExec(corePath string) bool {
...
 	}
 
-	// If the core snap doesn't support re-exec or run-from-core then don't do it.
-	if !coreSupportsReExec(corePath) {
+	if !strings.HasPrefix(exe, dirs.SnapMountDir) {
 		return distroTool
 	}
 
-	return coreTool
+	// if we are re-execed, then snap-seccomp is at the same location
```
Comment (zyga) [inline cmd/cmd.go]: I'd not mention snap-seccomp specifically here, it's a generic function

File: cmd/cmd_test.go [hunk 5]
```diff
@@ -196,67 +197,27 @@ func (s *cmdSuite) TestCoreSupportsReExec(c *C) {
...
-	// but, disabled by env
-	os.Setenv("SNAP_REEXEC", "0")
-	defer os.Unsetenv("SNAP_REEXEC")
+func (s *cmdSuite) TestInternalToolPathNoReexec(c *C) {
+	restore := cmd.MockOsReadlink(func(string) (string, error) {
+		return "/usr/lib/snapd/snapd", nil
+	})
+	defer restore()
 
 	c.Check(cmd.InternalToolPath("potato"), Equals, filepath.Join(dirs.DistroLibExecDir, "potato"))
```
Comment (zyga) [inline cmd/cmd_test.go]: This will not work on fedora, can you please use DistroLibExecDir above please?

**Pull request merged**